### PR TITLE
Neural Network segmentation

### PR DESCRIPTION
**Pull request by ebretl**
This code accomplishes:
 - labeling images or folders of images with polygons (LabelMe)
 - training a small semantic segmentation neural network
 - a line detector backed by a pre-trained neural network
 - all labeling and training scripts are run via roslaunch

### CODE REVIEW

File: rr_ml/nodes/image_label_tools/extract_images.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import os
+import random
+
+import rospy
+import rosbag
+from cv_bridge import CvBridge
+import cv2
+
+from . import util
+
```
Comment (chsahit) [inline rr_ml/nodes/image_label_tools/extract_images.py]: It's an important caveat that this only extracts a random subset of the images in the bag (unless you set num_images to be the number of images in the bag), I would add that in as a comment somewhere

File: rr_ml/nodes/semantic_segmentation/gen_fake_data.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    local_noise_mat = np.random.randint(local_noise_range[0], local_noise_range[1], img.shape)
+    global_noise_mat = np.ones(img.shape, dtype=np.int16) * np.random.randint(*global_noise_range)
+    img = (img + local_noise_mat + global_noise_mat).astype(np.uint8)
+
+    if not os.path.exists(data_dir):
+        os.makedirs(data_dir)
+
+    cv2.imwrite(os.path.join(data_dir, "input_%d.png" % n), img)
+    np.save(os.path.join(data_dir, "label_%d.npy" % n), label)
```
Comment (chsahit) [inline rr_ml/nodes/semantic_segmentation/gen_fake_data.py]: Doesn't Keras also have functionality for augmenting data, why don't you use that? or some combination of this custom augmentation code as well as what keras provides?

File: rr_ml/nodes/semantic_segmentation/model_utils.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        self.in_width = 256
+
+        self.orig_width = None
+        self.orig_height = None
+
+        self.n_channels = 5
+        self.n_classes = 2
+
+    # significant credit to https://github.com/divamgupta/image-segmentation-keras
+    def make_unet_model(self):
```
Comment (chsahit) [inline rr_ml/nodes/semantic_segmentation/model_utils.py]: surprised I don't see any batchnorm in this architecture, but if this works then I guess you're fine.

File: rr_ml/nodes/semantic_segmentation/run.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+        img = unet_helper.crop(cam_img)
+        cropped_height, cam_width = img.shape[:2]
+
+        X = unet_helper.prepare_input(img)
+
+        t0 = time.time()
+        y = model.predict_on_batch(X[np.newaxis])[0]
+        t1 = time.time()
+        print "neural net latency:", (t1 - t0)
```
Comment (chsahit) [inline rr_ml/nodes/semantic_segmentation/run.py]: maybe consider doing this as a moving average if you're going to consider using latency as a metric to measure. If you no longer need/use latency information, maybe print it out with ros_debug so you're not getting too much output.

File: rr_ml/nodes/image_label_tools/label_image.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import os
+
+import rospy
+from . import util
+
+osp = os.path
```
Comment (chsahit) [inline rr_ml/nodes/image_label_tools/label_image.py]: you're not using this variable in this file

File: rr_ml/nodes/image_label_tools/util.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import subprocess
+import json
+import tempfile
+
+import labelme
+import cv2
+import numpy as np
+import yaml
+
+osp = os.path
```
Comment (chsahit) [inline rr_ml/nodes/image_label_tools/util.py]: you're not using osp in this file either!

File: rr_ml/nodes/semantic_segmentation/run.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    rospy.Subscriber(image_topic, Image, image_msg_callback, queue_size=1, buff_size=10**8)
+
+    unet_helper = model_utils.UNetModelUtils()
+    cv_bridge = CvBridge()
+
+    model = keras.models.load_model(path_to_model)
+
+    while not rospy.is_shutdown():
+        if last_msg is None:
+            time.sleep(0.01)
```
Comment (chsahit) [inline rr_ml/nodes/semantic_segmentation/run.py]: consider making 0.01 a rosparam?

File: rr_ml/nodes/semantic_segmentation/train.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        yield inputs, labels
+
+
+def data_gen_wrapper(helper, input_files, label_files, epochs, do_augmentation):
+    for epoch in range(epochs):
+        for x in data_gen(helper, input_files, label_files, do_augmentation):
+            yield x
+
+
+def focal_loss(gamma):
```
Comment (chsahit) [inline rr_ml/nodes/semantic_segmentation/train.py]: a comment sitting above this function defining focal_loss would be nice

**Comment by ebretl** [comment]
I think we should just merge this in. It's still kinda cancer to work with because few things are properly documented. However, once it's in master we can put different aspects of how to clean up / streamline this code into our issue tracker

File: rr_ml/readme.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+`sudo python -m pip install keras tensorflow-gpu numpy labelme pyyaml` \
+OR \
+`python -m pip install --user keras tensorflow-gpu numpy labelme pyyaml`
+
+
+## Image Labeling Pipeline
+
+1. Extract image data from bag file: 
+`roslaunch rr_ml extract_images_iarrc_lines.launch bag_file:=/path/to/bag image_topic:=/topic num_images:=30`
```
Comment (Daniel-Martin576) [inline rr_ml/readme.md]: Not working when I try it...
```roslaunch rr_ml extract_images_iarrc_lines.launch bag_file:=/home/dani/Downloads/iarrc19_full_run.bag image_topic:=/camera_center/img_color_rect num_images:=30```
Not sure what is breaking.

**Comment by Daniel-Martin576** [comment]
I was reading through the code yesterday and I think we should merge it as it is. There are some things that maybe should be changed/cleaned but it definitely works and we can fix it more once it is in master.

**Comment by Daniel-Martin576** [comment]
I was getting around 80 ms just for the predicting which is a bit slow. Not sure if it is because the NN architecture you chose or what, but we can look at it later.

**Comment by Daniel-Martin576** [comment]
Feel free to make a new issue once you merge it of how to potentially improve it.

**Comment by Daniel-Martin576** [comment]
Not sure where all the labeled images you have are, but it might be beneficial if you posted them somewhere for future members to train the model with.

**Comment by ebretl** [comment]
@Daniel-Martin576 @dsiegler2000 either of you can merge this. I forgot to hit the merge button 2 months ago and now my write access to the repo has been rightfully revoked :sad:
